File: vigenere/Python/vigenere/tests/test_vigenerecipher.py
import unittest
from vigenere import vigencrypt, vigdecrypt

class VigenereCipherTests(unittest.TestCase):

    # I am aware that the brief says to include appropriate unit tests for all your functions.
    # However, it also says to expose functions vigencrypt and vigdecrypt to the user. Unlike in R,
    # it is not possible to include tests for generate_key_string, letter_units and shift_n_units
    # without exposing them to the user. Therefore, I have just included tests for vigencrypt
    # and vigdecrypt. It should be noted that these tests for vigencrypt and vigdecrypt encapsulate
    # the tests for generate_key_string, letter_units and shift_n_units that I have included in Q1;
    # I only included them in Q1 for completeness.

    # tests for vigencrypt and vigdecrypt
    
    def test_vigencrypt_example(self):
        '''Test that vigencrypt correctly encrypts the message imperialcollege with key
        word of MATRIX as given in the brief.'''
        self.assertEqual(vigencrypt('imperialcollege', 'MATRIX'), 'UMIVZFMLVFTIQGX')

    def test_vigdecrypt_example(self):
        '''Test that vigdecrypt correctly decrypts the message UMIVZFMLVFTIQGX with key
        word of MATRIX as given in the brief.'''
        self.assertEqual(vigdecrypt('UMIVZFMLVFTIQGX', 'MATRIX'), 'imperialcollege')

    def test_vigencrypt_different_cases(self):
        '''Test that supplying the message and key word to vigencrypt in oppisite cases
        to those indicated in the brief rasies an error.'''
        self.assertRaises(ValueError, vigencrypt, 'IMPERIALCOLLEGE', 'matrix')

    def test_vigdecrypt_different_cases(self):
        '''Test that supplying the message and key word to vigdecrypt in oppisite cases
        to those indicated in the brief rasies an error.'''
        self.assertRaises(ValueError, vigdecrypt, 'umivzfmlvftiqgx', 'matrix')

    def test_vigencrypt_non_alphabetic_message(self):
        '''Test that supplying a message to vigencrypt containing non-alphabetic characters
        rasies an error.'''
        self.assertRaises(ValueError, vigencrypt, '1@p', 'MATRIX')

    def test_vigdecrypt_empty_key(self):
        '''Test that supplying an empty key word to vigdecrypt rasies an error.'''
        self.assertRaises(ValueError, vigdecrypt, 'imperialcollege', '')

File: vigenere/R/vigenere/tests/testthat/test-vigenerecipher.R
# tests for generate_key_string, letter_units and shift_n_units

test_that('generate_key_string correctly computes the key string when the
          message is imperialcollege and the keyword is MATRIX, as given in
          the brief', {
  expect_equal(generate_key_string('imperialcollege', 'MATRIX'), 'MATRIXMATRIXMAT')
})

test_that('letter_units correctly computes the letter units of A', {
  expect_equal(letter_units('A'), 0)
})

test_that('shift_n_units correctly computes A shifted by 1 unit', {
  expect_equal(shift_n_units('A', 1), 'B')
})

# tests for vigencrypt and vigdecrypt

test_that('vigencrypt correctly encrypts the message imperialcollege with
          keyword of MATRIX, as given in the brief', {
  expect_equal(vigencrypt('imperialcollege', 'MATRIX'), 'UMIVZFMLVFTIQGX')
})

test_that('vigdecrypt correctly decrypts the message UMIVZFMLVFTIQGX with
        keyword of MATRIX, as given in the brief', {
  expect_equal(vigdecrypt('UMIVZFMLVFTIQGX', 'MATRIX'), 'imperialcollege')
})

test_that('supplying a message and a keyword to vigencrypt in oppisite cases
        to those indicated in the brief rasies an error', {
  expect_error(vigencrypt('IMPERIALCOLLEGE', 'matrix'), 'Invalid message, Invalid key')
})

test_that('supplying a message and a keyword to vigdecrypt in oppisite cases
        to those indicated in the brief rasies an error', {
  expect_error(vigdecrypt('umivzfmlvftiqgx', 'matrix'), 'Invalid message, Invalid key')
})

test_that('supplying a message to vigencrypt containing non-alphabetic
          characters rasies an error', {
  expect_error(vigencrypt('1@p', 'MATRIX'), 'Invalid message')
})

test_that('supplying an empty keyword to vigdecrypt rasies an error', {
  expect_error(vigdecrypt('imperialcollege', ''), 'Invalid key')
})
